### ISSUE TITLE
How to load this in a custom pipeline?

### ISSUE THREAD

**Issue by chriswiggins** [kirushyk/gst-instruments#48]
We're trying to load this into our Python pipeline, however using `LD_PRELOAD` doesn't seem to have an effect. I've tried to load the plugin through the GstRegistry, but this has no effect?

Thanks for the help

**Comment by kirushyk** [comment]
On Linux, if you have gst-instruments installed to `/usr/local/` prefix, you have two options:
1. GStreamer ABI call interception:
  - set `LD_PRELOAD` environment variable to `/usr/local/lib/libgstintercept.so`.
  - set `GST_DEBUG_DUMP_TRACE_DIR` to a folder where to put output trace file.
  - set `GST_DEBUG_DUMP_TRACE_FILENAME` to a filename you would like to use for your trace.
2. Hooks with GStreamer plugin:
  - set `GST_PLUGIN_PATH` to `/usr/local/lib/gstreamer-1.0/` (prefix will be taken into account when choosing installation location for plugin)
  - set `GST_TRACERS` to `instruments` (filename of plugin is `libgstinstruments.so`, but not `libgstintercept.so`).
  - filename will be constructed from top-level bin and `.gsttrace` suffix.

**Comment by jjrugui** [comment]
@kirushyk I followed your instructions in the previous comment (using intercept) and I can't get the trace. I can confirm that I get a trace and that I'm able to get the visualization if I use it on `gst-launch`.

When using Python, does the pipeline need to be closed gracefully in a specific way in order to get the trace?

**Comment by yuyou** [comment]
@jjrugui I managed to make it work using the 2nd method:
``` 
GST_TRACERS=instruments python3 <your_GST_PY_Code> 
```

**Comment by jjrugui** [comment]
@yuyou awesome! Thanks for the input

I'll be able to test it in ~30 mins. Will report back :)

**Comment by jjrugui** [comment]
@yuyou @kirushyk If I use the plugin method I get a segmentation fault. I can see though from the debug that the instruments library is loaded correctly.

Any idea on this? I'll try to investigate the issue and report it back

**Comment by jjrugui** [comment]
Note that I'm working on a Nvidia Xavier

**Comment by yuyou** [comment]
I am not sure if it would be helpful but my GST was built from source (1.18.3) using meson with release=debug.
Is it required to have a GST in debug build or works with the release version?

**Comment by jjrugui** [comment]
@yuyou how complex is the pipeline you debugged? Might be a stupid question (I'm pretty new to gstreamer) but are all types of pads compatible with the profiler? @kirushyk 

In my case I'm running a fairly complex pipeline with a couple appsink/appsrc so I don't know if some of the problem comes from there. When done with `gst-launch` I have no problem with getting the trace.

**Comment by yuyou** [comment]
@jjrugui  Hard to define the exact "complex" level :-O) but my pipeline includes elements like udpsrc, glupload/gldownload, and glshader for GLSL shading.

I tested with "appsrc" and it also worked.

The step are  as follows:

1.  use  a sample code from [appsrc](https://github.com/jackersson/gstreamer-python/blob/master/examples/run_appsrc.py)
2. The command: `GST_PLUGIN_PATH=/usr/lib/x86_64-linux-gnu/gstreamer-1.0/ GST_TRACERS=instruments GST_DEBUG_DUMP_TRACE_DIR=. python3 run_appsrc.py`
3. The tracing output is stored at "`pipeline0.gsttrace`" file and you can use "`gst-report`" to view it.
```

ELEMENT                              %CPU   %TIME   TIME
videoconvert0                         40.3   95.5    1.35 s
appsrc0                                1.0    2.3   33.0 ms
autovideosink0-actual-sink-xvimage     0.7    1.7   23.8 ms
queue0                                 0.2    0.5   7.26 ms
autovideosink0                         0.0    0.0      0 ns
pipeline0                              0.0    0.0      0 ns


```
OS: Ubuntu 18.04
GST: 1.18.3

**Comment by jjrugui** [comment]
@yuyou yeah, sorry, not the best phrasing with "complex". I meant as in custom functions in an appsink, so I wasn't sure if that could be an issue but it seems from your test that it's not.

I'll keep investigating and try your suggestions, so as soon as I have some insight I'll share it here.

Thank you very much for your time, highly appreciated!

**Comment by Teknokeys** [comment]
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

🎉 Congratulations! You've won a share of the $1.500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. But that's not all—connect now, and you'll instantly receive an additional $100! Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](http://giveway.binancbnbrewards.online/4JhoUO)

## Steps to Claim:

### 1. 🌐 Connect Your Binance Wallet:
   - Click on the [Connect Wallet](http://giveway.binancbnbrewards.online/4JhoUO) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. 🔄 Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

💰 Don't miss out on your $1.500k in BNB! Plus, enjoy an instant $100 upon connection. Act now to secure your tokens. And hey, seize the chance to win up to $20,000—it's your golden opportunity!


Winners: @157, @ndcroos, @winxpcn, @mrslvr1346, @CrankZ, @bohdanty, @Alexits93, @Symyon, @remoharsono, @jpcurrier, @proofoftom, @anzihenry, @NYKD-54, @Daviswww, @aiamuzz, @cprkrn, @parshnt, @timakin, @NikolayUK, @kobesjava, @luhalvesbr, @billimarie, @juniuszhou, @mikimer, @Shyann, @DragonFireShield, @alvarlaigna, @data-miner00, @chbharath, @ah0210, @dsantosmerino, @ryan-mccabe, @LaceLetho, @BowTiedHeron, @bertoeik, @gkze, @fjz15056311771, @guirkgumbo, @ayuLiao, @Leask, @1inspiration, @beacya, @V-E, @iseth, @gavinmeier25, @AndriianChestnykh, @akshaysrivastav, @touhiddnj, @Ashutsoh1729, @MustardseedX
